pkg/debug/common/common.go

```diff
@@ -0,0 +1,317 @@
+// Copyright 2019 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/remotecommand"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+
+	kubectlscheme "k8s.io/kubectl/pkg/scheme"
+)
+
+var (
+	ColorReset = "\033[0m"
+	ColorRed   = "\033[31m"
+	ColorGreen = "\033[32m"
+	ColorCyan  = "\033[36m"
+	scheme     = runtime.NewScheme()
+)
+
+type PodName struct {
+	PodName              string
+	PodNamespace         string
+	ChaosDaemonName      string
+	ChaosDaemonNamespace string
+}
+
+func init() {
+	_ = v1alpha1.AddToScheme(scheme)
+	_ = clientgoscheme.AddToScheme(scheme)
+}
+
+func UpperCaseChaos(str string) string {
+	parts := regexp.MustCompile("(.*)(chaos)").FindStringSubmatch(str)
+	return strings.Title(parts[1]) + strings.Title(parts[2])
+}
+
+// ExtractFromJson extract certain item from given string slice
+// String means the key of yaml
+// Number means the order in a slice
+func ExtractFromJson(chaos runtime.Object, str []string) (interface{}, error) {
```

Comment:
Do you need something like `reflect` :rofl: ?

Reply:
Thx for the reminder! That's what I need when I was using `kubectl` for dealing with different types of return values. I just need to return string for now.

Reply:
I mean... `reflect` could be a better way to get a string through a path (like '.Spec.Action') of struct (than serialize and deserialize into map).

If it's needed (for example, this deep field is common for a lot of chaos), you could add it to `chaos-builder` code generator. It's much safer than both `reflect` and `ExtractFromJson`

Reply:
Sorry I couldn't find how to use `reflect` to deal with this job... But I still managed to remove `ExtractFromJson`, by adding a type assertion part in each `Debug` function for different chaos, just like how controllers do. And after that I could get a string through a path.

BTW, it seems that I could not use getter in this circumstance. Since I could not type assert `runtime.Object` (which is what I get from `AllKinds`) to a common chaos type like `InnerObject`, I still have to do type assertion to specific chaos to get what I need. Is there someplace I misunderstood?

Reply:
> I could not type assert `runtime.Object` (which is what I get from AllKinds) to a common chaos type like `InnerObject`

Why you cannot? In `pkg/router`, the `Reconcile` method for `Reconciler` asserts `runtime.Object` into `v1alpha1.InnerSchedulerObject`. If the object do have these methods, I think the type conversion is available.

Reply:
My bad, I used type assertion in the wrong way for it.